pkg/kv/kvserver/replica_raftstorage.go

```diff
@@ -284,16 +284,18 @@ func (r *Replica) GetTerm(i uint64) (uint64, error) {
 func term(
 	ctx context.Context,
 	rsl stateloader.StateLoader,
-	reader storage.Reader,
+	eng storage.Engine,
 	rangeID roachpb.RangeID,
 	eCache *raftentry.Cache,
 	i uint64,
 ) (uint64, error) {
 	// entries() accepts a `nil` sideloaded storage and will skip inlining of
 	// sideloaded entries. We only need the term, so this is what we do.
-	ents, err := entries(ctx, rsl, reader, rangeID, eCache, nil /* sideloaded */, i, i+1, math.MaxUint64 /* maxBytes */)
+	ents, err := entries(ctx, rsl, eng, rangeID, eCache, nil /* sideloaded */, i, i+1, math.MaxUint64 /* maxBytes */)
 	if errors.Is(err, raft.ErrCompacted) {
-		ts, err := rsl.LoadRaftTruncatedState(ctx, reader)
+		// TODO(pavelkalinnikov): maybe push this check to entries(), because it
```

Comment:
Seemingly, the only thing that `term` and `entries` do differently here is:
- `term` loads the truncated state unconditionally, because it would contain the term of the last compacted entry
- `entries` only loads it when it could not read any entries

Reply:
The code in `entries` seems a bit more complicated than it needs to be. On the face of it this could be explained by a desire to assert something. But really the assertions are quite weird; note how we maintain `expectedIndex` but (in `scanFunc`) silently bail out if we find a gap. I think *that* is the one place where we should fail if we see a gap. 
Then, when the scan is done, we should check

- if there was an entry, check that it matches `lo`. If so,
- check if we had the number of entries requested or we hit the max bytes limit. If so,
- we're good.
- Otherwise, load first and return ErrCompacted if lo < the first index.
- Otherwise, load last index and return ErrUnavailable if hi > lastIndex.
- otherwise, return an assertion failed error.

I agree that we should be able to do the code movement you suggested now. Perhaps we can clean up `entries` too as per the above, either in a separate commit, or PR.

Reply:
I decoupled `term` and `entries`, so this TODO is no longer a concern.

Will look at refactoring `entries` in a follow-up PR.